### PR TITLE
PoC: Add lifecycle hooks, packageWillResolve

### PR DESCRIPTION
**Pull request by andreypopp**
**Do not merge:** This is for discussion.

This allows to override the default pnpm's package resolution algorithm. The
idea is pretty basic — we pass a set of lifecycle hooks to API calls which are
then processed by pnpm core algo.

The only hook I added now is

    packageWillResolve(spec: PackageSpec, opts: ResolveOptions): Promise<ResolveResult>

which can be used like:

    api.installPkgs('@opam/lwt@^2.5.0', {
      lifecycle: {
        packageWillResolve: async (spec, opts) => {
          if (
            ['range', 'version', 'tag'].indexOf(spec.type) > -1 &&
            spec.scope === '@opam'
          ) {
            return resolveFromOpam(spec, opts)
          } else {
            return null
          }
        }
      }
    })

Related to #559

### CODE REVIEW

File: src/install/fetchResolution.ts [hunk 1]
```diff
@@ -6,6 +6,9 @@ import execa = require('execa')
 const gitLogger = logger('git')
 
 export default async function fetchRes (res: ResolveResult, target: string, opts: FetchOptions): Promise<void> {
+  if (res.fetch) {
+      return res.fetch(target)
+  }
```
Comment (andreypopp) [inline src/install/fetchResolution.ts]: This is so custom resolvers can override fetch and still specify `tarball` and `repo`.

File: src/api/install.ts [hunk 2]
```diff
@@ -35,6 +35,7 @@ export type InstallContext = {
   installationSequence: string[],
   fetchLocks: CachedPromises<void>,
   graph: Graph,
+  lifecycle: LifecycleHooks,
```
Comment (andreypopp) [inline src/api/install.ts:39]: Not sure if `InstallContext` is right place for this, suggestions?

File: src/types.ts [hunk 3]
```diff
@@ -1,4 +1,26 @@
...
+   * default resolution algorithm will be used.
+   */
+  packageWillResolve?: (spec: PackageSpec, opts: ResolveOptions) => Promise<ResolveResult | null>,
+
+  // TODO: add more lifecycle hooks
+  //
+  // packageDidResolve
+  //
+  // packageWillFetch
+  // packageDidFetch
```
Comment (andreypopp) [inline src/types.ts]: `packageDidFetch(spec: PackageSpec, target: string)` probably can be used to check if `package.json` is exists and generate one if needed.

**Comment by zkochan** [comment]
When I've read `lifecycle` my first thought was the script tasks, like install, postinstall, prepublish, etc.

I think stage would be better naming for this, or something else

**Comment by andreypopp** [comment]
> When I've read lifecycle my first thought was the script tasks, like install, postinstall, prepublish, etc.

Good point. What do you think of names of hooks: `packageDid*`/`packageWill*`. The alternative would be `onBefore*`/`on*`, for example: `onBeforeResolve`, `onResolveComplete`.

**Comment by zkochan** [comment]
`onSomething` sounds to me somewhat passive, like you can't change the behavior, just inspect. `pre/post Something` or your first choice might be better.

**Comment by zkochan** [comment]
So you idea is to rewrite even the standard specs, because your hook is executed before all the standard ones... `@opam/lwt@^2.5.0` is a standard spec. Do you think it is a good idea to rewrite the standard behavior? Maybe hooks should be for unhandled situations and nonstandard specs. Like `@opam/lwt@sometype:^2.5.0`

**Comment by andreypopp** [comment]
> Do you think it is a good idea to rewrite the standard behavior? 

Given that I want to extend installation in a non-standard ways — yes, in my case I want to assume that some scope is a *built-in* — `@opam/*` which fetches tarballs from OPAM registry and generates `package.json` on the fly. I think the same technique can be used to bridge any other package registry.

**Comment by andreypopp** [comment]
> When I've read lifecycle my first thought was the script tasks, like install, postinstall, prepublish, etc.

On a second thought — hooks introduced in this PR are similar to what declared in `package.json`. The difference are:

* they are declared not by package itself but by installer which extends pnpm
* they are more pnpm specific 

I even think we can add more hooks which are defined around hooks declared in `package.json`.

**Comment by zkochan** [comment]
So you want to use pnpm as a package manager for ocaml. But ocaml doesn't need the node_modules dependency structure. So you'll override the install as well?

**Comment by andreypopp** [comment]
> So you want to use pnpm as a package manager for ocaml. But ocaml doesn't need the node_modules dependency structure. So you'll override the install as well?

We already have a build system for OCaml (Esy) which processes dependencies from `node_modules/`. It works well with npm and I'm looking into pnpm as a more faster installer and also for an ability to bridge another registry of OCaml packages directly with npm workflow.

**Comment by andreypopp** [comment]
Rebased on the current master, also add new hooks.

**Comment by andreypopp** [comment]
Closing this as I don't think this worth pursuing. The better way would be to have a way to re-create installation algo using high-level ops.

**Pull request closed without merging**